### PR TITLE
Robustness metrics

### LINKED ISSUE #497: Document and implement one robustness metric for ensemble statistics
### Description

Look at how AR5 WG1 is estimating the confidence of the climate change signal from an ensemble and implement it.
See Box 12.1 in AR5 WG1 Chapter 12 for a discussion on this topic. 

The use case would be to reproduce figures where stippling/hatching on a map indicates model agreement. 

### TODO

- List the various algorithms used and estimate implementation time; 
- Discuss in group which ones we want to include in xclim;
- Discuss API;
- Submit PR.

### PR DESCRIPTION
**Pull request by aulemahal**
### Pull Request Checklist:
- [x] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #497
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Documentation has been added / updated (for bug fixes / features)
- [x] HISTORY.rst has been updated (with summary of main changes)
- [x] `bumpversion (minor / major / patch)` has been called on this branch
- [x] Tags have been pushed (`git push --tags`)

* **What kind of change does this PR introduce?** 
Adds the `xclim.robustness` module with its `robustness` function. It takes in the ensemble distribution of an indicator and returns a categorical map qualifying the level of agreement between members. Implementation is following articles listed in box 12.1 of Chapter 12 of the Working Group 1 AR5 Report (IPCC). 

 This PR will implement method "C", Knutti and Sedlacek (2013), which returns a floating robustness metric R and method "D", Tebaldi et al. (2011), which returns a mapping with values for "no significant change", "change but no agreement on sign", "change and agreement on sign". 

In the perspective of expanding the available methods, the submodule is structured similarly to `xclim.analog` : a master function takes DataArrays and a method name as input and then runs a apply_ufunc call for the given method, vectorizing it along the spatial dimensions. Metrics are defined as functions accepting numpy arrays and wrapped with the `metric` decorator that does some overhead work and registers the method.

* **Does this PR introduce a breaking change?** 
No.

* **Other information**:
Translating scientific articles to python functions has proven quite difficult here. If someone can compare the code with the article, I would be grateful. I haven't found other implementations of these metrics elsewhere to compare.

### CODE REVIEW

**Comment by aulemahal** [comment]
Thinking that would better organize the code, I moved the "ensembles" module to a folder, dividing the methods in to 3 files: "base", with `create_ensemble` and the ensemble statistics, "reduce", with the kkz and kmeans ensemble reducing methods and "robustness" with the new ensemble robustness metrics.

Also, added tests are a bit fragile. For the "knutti_sedlacek" metric I tested my code against itself, so yeah.

**Review by huard — COMMENTED** [review]
Haven't gone in details yet. Have you checked climpred ? They may have an implementation of the ranked skill score to guide you.

**Comment by huard** [comment]
@martinleduc @RondeauG Est-ce que vous avez déjà utilisé ce genre de métrique ? Êtes-vous capable de valider ce PR ?

**Comment by RondeauG** [comment]
Possibly ? I would have to read the paper to see how it compares to what I'm currently implementing (based on Sanderson, Knutti & Caldwell 2015). Given that Knutti is involved in both, there are probably parallels to be done. I can look at it tomorrow or next week.

File: xclim/ensembles/_robustness.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        metrics[func.__name__] = Metric(
+            _metric_overhead, mapping, type(na_value), hist_dims
+        )
+        return _metric_overhead
+
+    return _metric
+
+
+@metric(na_value=np.nan, hist_dims=["time"])
+def knutti_sedlacek(hist, sims):
```
Comment (RondeauG) [inline xclim/ensembles/_robustness.py:122]: This one looks fine

**Review by RondeauG — CHANGES_REQUESTED** [review]

File: xclim/ensembles/_robustness.py [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+    # When p < p_change, the hypothesis of no significant change is rejected.
+    # We need at least X % models showing significant change
+    if (p < p_change).sum() / hist.size <= X:
+        return 0  # No significant change
+
+    # Test that models agree on the sign of the change
+    change_sign = np.sign(sims.mean(axis=1) - hist).clip(0, 1)
+    if (1 - Y) <= change_sign.sum() / hist.size <= Y:
+        return 1  # Significant change but no agreement on sign of change
+    return 2  # Significant change and agreement on sign of change
```
Comment (RondeauG) [inline xclim/ensembles/_robustness.py:240]: Thinking about it, `change_sign.sum() / hist.size` itself would be a very useful output for the function. This could allow the user to create their own thresholds of significance (for example: Y>0.8 = "significative", Y>0.95 = "very significative"), instead of the binary `1` and `2` results.

**Comment by cothel** [comment]
Allô, 
Je suis en train de faire des choses sur ce sujet pour les vents dans le projets éolien à partir des méthodologies cités dans plusieurs études CORDEX. Dans mon cas, la conclusion serait que les membres de l'ensemble s'entendent sur l'insignifiance du changement. Je suis trop dans le jus (finir le dit projet éolien!) pour regarder le code, mais je sais déjà qu'il faut préalablement détecter un changement significatif dans un certain pourcentage des membres. Si je n'est pas dans les références citées, nous allons publier sur ça (j'en discute surtout avec @martinleduc ). Je vous envoie quelques diapos.
Il devrait y avoir de la flexibilité sur:
- le choix du test statistique
- choix du niveau de signification
- le choix du % des membres qui ont détecté un changement significatif
-le choix du % de membres de l'ensemble pour le niveau de consensus sur le signe du changement
[WEC2100_steering_commitee_20200505_HCote_extrait.pdf](https://github.com/Ouranosinc/xclim/files/5585776/WEC2100_steering_commitee_20200505_HCote_extrait.pdf)

**Comment by aulemahal** [comment]
Last commit is a draft of a refactor following reviews.
My suggestion is to divide `ensemble_robustness` in two:

- `ensemble_robustness` : Calculate the fraction of members showing significant change and the fraction of those members agreeing on the sign of change. Takes in `ref` and `fut`, the reference and future period values of the given indicator, both along `realization` and `time`. Also takes the name of the statistical test used for testing significant change (as the  `stat` argument) and kwargs related to that test. Returns a dataset with two variables : `change_frac` and `sign_frac`.
Right now, the `tebaldi_et_al` method as been translated to `stat='ttest'`. I'm understanding that what @RondeauG  calls the "IPCC" method, would simply ignore the "significant change" part.

- `robustness_map` : From the results of `ensemble_robustness` and user-provided threshold X and Y, returns a categorical map with integers 0, 1, 2. This one would thus be more a convenience method, given its simplicity. We could offer a possibility for having two Y thresholds, like in the IPCC reports.

`knutti_sedlacek` would be a third (standalone) method of the submodule.


From what I understand from this thread, this should cover most use cases? 

If we go back to box 12.1 of chapter 12 of the AR5 contribution of WG1 (or in short AR15WG1C12B12.1):

- Methods A and B need 500 years of pre-industrial runs to compute climate internal variability. I am guessing that this quite a niche thing to come by. However, we could let the user provide it's own estimation of  internal variability. But that doesn't quite plug into the suggest framework...
- Method E could also be implemented in the future. AFAIU, it would also be a second standalone method, because its categories are different than the suggested `robustness_map`. This is why I first suggested a function returning different _categories_, but I realize that this was too narrow in its generality.

File: xclim/core/options.py [hunk 3]
```diff
@@ -136,14 +136,15 @@ class set_options:
...
 
-    You can use ``set_options`` either as a context manager:
-    >>> import xclim
-    >>> import xarray
-    >>> ds  = xarray.open_dataset("path_to_tas_file").tas
-    >>> with xclim.set_options(metadata_locales=['fr'])
+    Examples
+    --------
+    # You can use ``set_options`` either as a context manager:
+    >>> import xclim as xc
```
Comment (Zeitsperre) [inline xclim/core/options.py]:
```suggestion
    >>> import xclim
```

File: xclim/ensembles/_reduce.py [hunk 4]
```diff
@@ -187,26 +187,26 @@ def kmeans_reduce_ensemble(
     Examples
     --------
     >>> from xclim.ensembles import create_ensemble, kmeans_reduce_ensemble
```
Comment (Zeitsperre) [inline xclim/ensembles/_reduce.py]:
```suggestion
    >>> import xclim
    >>> from xclim.ensembles import create_ensemble, kmeans_reduce_ensemble
```

**Comment by aulemahal** [comment]
Ok, so I abandoned the `robust_map` method, it seems so easy yet so customizable that I can't see a solid use case for a "hardcoded" xclim method. 

Also, `ensemble_robustness` was renamed to `change_significance` and it outputs 2 DataArrays instead of a dataset. I added Welch's T-test, as this is what @cothel seem to use and it's easy to add those.

Also added the `None`  method, that acts like all members show significant change and only computes the sign agreement fraction. It seems that @RondeauG, was doing that a some point?

I renamed `hist` to `ref` and `sims` to `fut`. 

Let me know what you think of all this. Sorry for the multiple refactoring...

File: xclim/ensembles/_robustness.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+            )  # Merge to get all "discontinuities" of the CDF difference
+            y1_f = np.insert(
+                y1, x2_in_1, np.r_[0, y1][x2_in_1]
+            )  # y1 with repeated value (to the right) where x2 is inserted
+            y2_f = np.insert(
+                y2, x1_in_2, np.r_[0, y2][x1_in_2]
+            )  # Same for y2. 0s are prepended where needed.
+
+            # Discrete integral of the squared difference (distance) between the two CDFs.
+            return np.sum(np.diff(x) * (y1_f - y2_f)[:-1] ** 2)
```
Comment (huard) [inline xclim/ensembles/_robustness.py:203]: Have you considered `np.trapz` for the integration ?
`scipy.stats.scoreatpercentile` for the cdf ?

Comment (huard) [reply xclim/ensembles/_robustness.py:203]: or the `quantile` method ?

File: xclim/ensembles/_robustness.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+      Other arguments specific to the statistical test.
+
+      For 'ttest' and 'welch-ttest':
+        p_change : float (default : 0.05)
+          p-value threshold for rejecting the hypothesis of no significant change.
+
+    Returns
+    -------
+    change_frac: DataArray
+      The fraction of members that show significant change. (0..1)
```
Comment (huard) [inline xclim/ensembles/_robustness.py:36]:
```suggestion
      The fraction of members that show significant change [0, 1].
```

**Review by huard — APPROVED** [review]

File: xclim/ensembles/_robustness.py (hunk 5 above)
Comment (aulemahal) [reply xclim/ensembles/_robustness.py:203]: I forgot about "trapz"!  I just tested and speed is pretty much the same as without it. 

As for the other two and AFAIU: What I find nice about this solution is that it is _exact_. If we define the experimental CDF as a stepwise function, this code gives the exact definite integral of the (CDF2 - CDF1)**2. No need to use fancy functions to get a smooth or uniformly-spaced array!

Comment (huard) [reply xclim/ensembles/_robustness.py:203]: Good for me.

**Comment by huard** [comment]
I think this is good to go !

**Review by RondeauG — APPROVED** [review]

**Comment by aulemahal** [comment]
Last commit made some fixes to the documentation : I moved the "ensembles" api higher in the API page, divided it between base/reduction/robustness and added a introduction paragraph to robustness with a reference to AR15WG1C12B12.1.
And bumped the version.

**Comment by aulemahal** [comment]
Adaptation to be able to pass `deltas` as first argument. Also, new test `threshold` for simple absolute or relative delta thresholds.

New test and example in the docstring.

This adds complexity and a few conditional cases in the docstring... I hope it is still clear.

**Review by huard — APPROVED** [review]

**Review by RondeauG — APPROVED** [review]

**Comment by aulemahal** [comment]
Peut-être c'est la fin?

I fixed some edge cases with the new nan policy. Also adapted tests and completely implemented accepting Datasets as inputs. It was already feasible, but with some edge cases. When using datasets, `fut` and `ref` must be of the same type and have the same variables, otherwise it's quite transparent.

              _      _      _
              >(.)__ <(.)__ =(.)__
               (___/  (___/  (___/

**Comment by huard** [comment]
Two tests are failing on my end: 
<details>
================================================================================================ FAILURES =================================================================================================
__________________________________________________________________________________ test_change_significance_delta[True] ___________________________________________________________________________________

robust_data = (<xarray.Dataset>
Dimensions:  (lon: 4, realization: 4, time: 40)
Dimensions without coordinates: lon, realization, ti..., time
Data variables:
    tas      (lon, realization, time) float64 dask.array<chunksize=(1, 4, 40), meta=np.ndarray>)

    def test_change_significance_delta(robust_data):
        ref, fut = robust_data
        delta = fut.mean("time") - ref.mean("time")
        chng, sign = ensembles.change_significance(delta, test="threshold", abs_thresh=2)
        if isinstance(ref, xr.Dataset):
            chng = chng.tas
            sign = sign.tas
>       np.testing.assert_array_equal(chng, [0, 0, 0.5, np.nan])

tests/test_ensembles.py:561: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

x = array([0. , 0. , 0.5, 0. ]), y = array([0. , 0. , 0.5, nan]), func = <ufunc 'isnan'>, hasval = 'nan'

    def func_assert_same_pos(x, y, func=isnan, hasval='nan'):
        """Handling nan/inf.
    
        Combine results of running func on x and y, checking that they are True
        at the same locations.
    
        """
        x_id = func(x)
        y_id = func(y)
        # We include work-arounds here to handle three types of slightly
        # pathological ndarray subclasses:
        # (1) all() on `masked` array scalars can return masked arrays, so we
        #     use != True
        # (2) __eq__ on some ndarray subclasses returns Python booleans
        #     instead of element-wise comparisons, so we cast to bool_() and
        #     use isinstance(..., bool) checks
        # (3) subclasses with bare-bones __array_function__ implementations may
        #     not implement np.all(), so favor using the .all() method
        # We are not committed to supporting such subclasses, but it's nice to
        # support them if possible.
        if bool_(x_id == y_id).all() != True:
            msg = build_err_msg([x, y],
                                err_msg + '\nx and y %s location mismatch:'
                                % (hasval), verbose=verbose, header=header,
                                names=('x', 'y'), precision=precision)
>           raise AssertionError(msg)
E           AssertionError: 
E           Arrays are not equal
E           
E           x and y nan location mismatch:
E            x: array([0. , 0. , 0.5, 0. ])
E            y: array([0. , 0. , 0.5, nan])

../../.conda/envs/xclim/lib/python3.6/site-packages/numpy/testing/_private/utils.py:740: AssertionError
__________________________________________________________________________________ test_change_significance_delta[False] __________________________________________________________________________________

robust_data = (<xarray.DataArray 'tas' (lon: 4, realization: 4, time: 40)>
array([[[274.58932717, 274.2998733 , 273.38842172, 273.62...     276.13121607, 272.20878628, 274.93243988, 274.99455816]]])
Dimensions without coordinates: lon, realization, time)

    def test_change_significance_delta(robust_data):
        ref, fut = robust_data
        delta = fut.mean("time") - ref.mean("time")
        chng, sign = ensembles.change_significance(delta, test="threshold", abs_thresh=2)
        if isinstance(ref, xr.Dataset):
            chng = chng.tas
            sign = sign.tas
>       np.testing.assert_array_equal(chng, [0, 0, 0.5, np.nan])

tests/test_ensembles.py:561: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

x = array([0. , 0. , 0.5, 0. ]), y = array([0. , 0. , 0.5, nan]), func = <ufunc 'isnan'>, hasval = 'nan'

    def func_assert_same_pos(x, y, func=isnan, hasval='nan'):
        """Handling nan/inf.
    
        Combine results of running func on x and y, checking that they are True
        at the same locations.
    
        """
        x_id = func(x)
        y_id = func(y)
        # We include work-arounds here to handle three types of slightly
        # pathological ndarray subclasses:
        # (1) all() on `masked` array scalars can return masked arrays, so we
        #     use != True
        # (2) __eq__ on some ndarray subclasses returns Python booleans
        #     instead of element-wise comparisons, so we cast to bool_() and
        #     use isinstance(..., bool) checks
        # (3) subclasses with bare-bones __array_function__ implementations may
        #     not implement np.all(), so favor using the .all() method
        # We are not committed to supporting such subclasses, but it's nice to
        # support them if possible.
        if bool_(x_id == y_id).all() != True:
            msg = build_err_msg([x, y],
                                err_msg + '\nx and y %s location mismatch:'
                                % (hasval), verbose=verbose, header=header,
                                names=('x', 'y'), precision=precision)
>           raise AssertionError(msg)
E           AssertionError: 
E           Arrays are not equal
E           
E           x and y nan location mismatch:
E            x: array([0. , 0. , 0.5, 0. ])
E            y: array([0. , 0. , 0.5, nan])

../../.conda/envs/xclim/lib/python3.6/site-packages/numpy/testing/_private/utils.py:740: AssertionError
============================================================================================ warnings summary =============================================================================================

</details>
